### PR TITLE
Delegate workflow spec creation

### PR DESCRIPTION
**Pull request by nolag**

### CODE REVIEW

File: core/services/workflows/workflow_spec_factory.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	if !ok {
+		return sdk.WorkflowSpec{}, "", ErrInvalidWorkflowType
+	}
+
+	rawSpec, err := factory.GetRawSpec(workflow)
+	if err != nil {
+		return sdk.WorkflowSpec{}, "", err
+	}
+
+	spec, err := factory.GetSpec(rawSpec, config)
```
Comment (cedric-cordenier) [inline core/services/workflows/workflow_spec_factory.go:37]: @nolag Will we ever want to call GetSpec + GetRawSpec separately? If not I think it makes sense to combine the two functions into one.

**Review by cedric-cordenier — DISMISSED** [review]

File: core/scripts/go.mod [hunk 2]
```diff
@@ -379,3 +379,5 @@ replace (
 	// replicating the replace directive on cosmos SDK
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
 )
+
+exclude github.com/sourcegraph/sourcegraph/lib v0.0.0-20221216004406-749998a2ac74
```
Comment (nolag) [inline core/scripts/go.mod]: See https://github.com/smartcontractkit/chainlink/pull/14446, I merged to my local one so that I could get the IDE working.



That PR should merge before mine so the diff won't show up.

File: core/services/workflows/workflow_spec_factory.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/smartcontractkit/chainlink-common/pkg/workflows/sdk"
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+)
+
+var ErrInvalidWorkflowType = errors.New("invalid workflow type")
+
+type SDKWorkflowSpecFactory interface {
+	GetSpec(rawSpec, config []byte) (sdk.WorkflowSpec, error)
+	GetRawSpec(wf string) ([]byte, error)
```
Comment (nolag) [inline core/services/workflows/workflow_spec_factory.go:17]: Two functions will allow us to potentially have WASM files loaded in different ways.  I figured that EOD it will store something on the box, so there would be a file version.  There may also be one that resolves HTTP at runtime, I figured that could download embed the WASM file version, but override GetRawSpec to do the HTTP download first.

**Review by cedric-cordenier — DISMISSED** [review]

File: core/services/workflows/workflow_spec_factory.go (hunk 3 above)
Comment (nolag) [reply core/services/workflows/workflow_spec_factory.go:17]: Also, it could allow caching later if we watned.

File: core/services/job/models_test.go [hunk 4]
```diff
@@ -322,6 +324,9 @@ func TestWorkflowSpec_Validate(t *testing.T) {
 			w := &WorkflowSpec{
 				Workflow: tt.fields.Workflow,
 			}
+			sum := sha256.New()
+			sum.Write([]byte("test"))
+			w.WorkflowID = fmt.Sprintf("%x", sum.Sum(nil))
```
Comment (cedric-cordenier) [inline core/services/job/models_test.go]: I think you can remove this altogether; the only reason we set this is to pass the assertion on line 333 (because Validate sets this)

Both can go as it shouldn't be relevant

Comment (nolag) [reply core/services/job/models_test.go]: Validate fails if the size of the workflow id isn't right.  Wasn't sure if it was ok to remove that.

File: core/services/job/models.go [hunk 5]
```diff
@@ -886,7 +896,6 @@ func (w *WorkflowSpec) Validate() error {
 	}
 	w.WorkflowOwner = strings.TrimPrefix(s.Owner, "0x") // the json schema validation ensures it is a hex string with 0x prefix, but the database does not store the prefix
 	w.WorkflowName = s.Name
-	w.WorkflowID = s.CID()
```
Comment (bolekk) [inline core/services/job/models.go:889]: why is this removed?

File: core/services/job/models.go [hunk 6]
```diff
@@ -857,16 +858,25 @@ type LiquidityBalancerSpec struct {
...
-	WorkflowOwner string    `toml:"-" db:"workflow_owner"` // Derived. Do not modify. the owner of the workflow.
-	WorkflowName  string    `toml:"-" db:"workflow_name"`  // Derived. Do not modify. the name of the workflow.
-	CreatedAt     time.Time `toml:"-"`
-	UpdatedAt     time.Time `toml:"-"`
+	WorkflowID    string           `toml:"-" db:"workflow_id"`    // Derived. Do not modify. the CID of the workflow.
+	WorkflowOwner string           `toml:"-" db:"workflow_owner"` // Derived. Do not modify. the owner of the workflow.
+	WorkflowName  string           `toml:"-" db:"workflow_name"`  // Derived. Do not modify. the name of the workflow.
+	CreatedAt     time.Time        `toml:"-"`
+	UpdatedAt     time.Time        `toml:"-"`
+	SpecType      WorkflowSpecType `json:"spec_type"`
```
Comment (bolekk) [inline core/services/job/models.go]: Does this break anything? Is a default value set automatically?

File: core/services/workflows/engine.go [hunk 7]
```diff
@@ -853,7 +854,7 @@ func (e *Engine) Close() error {
 }
 
 type Config struct {
-	Spec                 string
```
Comment (bolekk) [inline core/services/workflows/engine.go:856]: "Spec" feels OK, why the change?

File: core/services/job/models.go (hunk 5 above)
Comment (nolag) [reply core/services/job/models.go:889]: The WASM shouldn't report the CID.  The host should be responsible for it.  @cedric-cordenier and I were talking, I'll move the calculation here.



I think I have more work to do.

File: core/services/workflows/engine.go (hunk 7 above)
Comment (nolag) [reply core/services/workflows/engine.go:856]: it wasn't used and spec is't always going to be a string.

File: core/services/job/models.go (hunk 6 above)
Comment (nolag) [reply core/services/job/models.go]: I defaulted to YAML if it's not set.

File: core/services/workflows/workflow_spec_factory.go (hunk 3 above)
Comment (nolag) [inline core/services/workflows/workflow_spec_factory.go]: needed to make CID

File: core/services/job/models.go [hunk 8]
```diff
@@ -895,6 +914,20 @@ func (w *WorkflowSpec) Validate() error {
 	return nil
 }
 
+func (w *WorkflowSpec) SdkWorkflowSpec(ctx context.Context) (sdk.WorkflowSpec, error) {
```
Comment (jmank88) [inline core/services/job/models.go]:
```suggestion
func (w *WorkflowSpec) SDKWorkflowSpec(ctx context.Context) (sdk.WorkflowSpec, error) {
```
Or just?
```suggestion
func (w *WorkflowSpec) SDKSpec(ctx context.Context) (sdk.WorkflowSpec, error) {
```

**Review by cedric-cordenier — DISMISSED** [review]

**Review by cedric-cordenier — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]
